pkg/trace: add new stats endpoint for client computed stats

gbbr opened the pull request.

furmmon commented on pkg/trace/agent/agent.go:
```diff
@@ -231,11 +232,13 @@ func (a *Agent) Process(p *api.Payload, sublayerCalculator *stats.SublayerCalcul
-			Sublayers: pt.Sublayers,
-			Env:       pt.Env,
-		})
+		if !p.ClientComputedStats {
```
We'll still want to send sublayer stats to the concentrator

gbbr replied:
Ok I think I got it now.

gbbr committed:
Remove extra channel

gbbr committed:
todo

piochelepiotr committed:
pkg/trace/stats/quantile: accept ddsketch distributions (#6857)

Update the stats endpoint of the agent to accept ddsketches.

We receive 2 ddsketches: one for errors, one for hits. They are serialized with protobuf inside the stats message.

furmmon commented on pkg/trace/agent/agent.go:
```diff
@@ -259,6 +263,91 @@ func (a *Agent) Process(p *api.Payload, sublayerCalculator *stats.SublayerCalcul
+				Name:     b.Name,
+				Measure:  stats.HITS,
+				TagSet:   tagset,
+				TopLevel: float64(b.Hits),
```
@piochelepiotr can we actually not set TopLevel at all?

piochelepiotr replied:
Yes, we can skip it.

gbbr replied:
Should I remove this completely or do it in another PR? Not sure what you mean here...

furmmon replied:
You can remove it entirely on this PR (scoped to the new stat payload). This field is not used anymore

gbbr replied:
Done.

furmmon commented on pkg/trace/agent/agent.go:
```diff
@@ -259,6 +263,91 @@ func (a *Agent) Process(p *api.Payload, sublayerCalculator *stats.SublayerCalcul
+			newb.Counts[key] = stats.Count{
+				Key:      key,
+				Name:     b.Name,
+				Measure:  stats.HITS,
```
I think hits and hitsLatency misalignment is confusing. Currently 
- hits includes errors
- hitsLatencies exclude errorLatencies

I think we should either rename hitsSummary to okHitsSummary or have both exclusive

gbbr replied:
I think `hits` is `hits` regardless of outcome. No need to rename that. However, for cases when `hits` does not include all _hits_ (excludes errors) then we should likely use the word "success" or "nonError". Then renaming to `nonErrorLatencies` is clear. WDYT?

furmmon commented on pkg/trace/api/api.go:
```diff
@@ -357,6 +367,48 @@ func (r *HTTPReceiver) rateLimited(n int64) bool {
+			return
+		}
+	default:
+		if err := proto.Unmarshal(slurp, &in); err != nil {
```
I think we should now cleanup proto support as it's implemented with msgpack

gbbr replied:
Done.

gbbr committed:
Remove proto

bmermet commented on pkg/trace/test/agent.go:
```diff
@@ -41,6 +41,9 @@ func newAgentRunner(ddAddr string, verbose bool) (*agentRunner, error) {
 		if verbose {
 			log.Print("agent: trace-agent not found, trying to install...")
 		}
+		// TODO(x): Do not use $GOBIN/trace-agent. Install a new one every time.
```
Isn't this addressed in https://github.com/DataDog/datadog-agent/pull/6906 ?

gbbr replied:
True. Still need to merge that.

bmermet commented on pkg/trace/api/api.go:
```diff
@@ -390,6 +400,39 @@ func (r *HTTPReceiver) rateLimited(n int64) bool {
+	defer timing.Since("datadog.trace_agent.receiver.stats_process_ms", time.Now())
+
+	rd := NewLimitedReader(req.Body, r.conf.MaxRequestBytes)
+	slurp, err := ioutil.ReadAll(rd)
```
Why read the data in a byte array to wrap it later in a bytes.NewReader?

gbbr replied:
I guess it's a leftover from the previous change where we had multiple supported codecs. I'll change it.

gbbr replied:
Done. I removed a test which makes sure the behaviour is correct if the payload is larger than 10MB. Realistically speaking that should technically be impossible since most payloads have 1-2 buckets and that helper code is already well tested. Should be good enough as far as I'm concerned.

gbbr commented on pkg/trace/agent/normalizer.go:
```diff
@@ -10,12 +10,8 @@ import (
 	"fmt"
 	"math"
 	"strconv"
-	"strings"
```
I don't think so. That refactoring doesn't make sense outside this PR. This change is what brings in the new requirement so the refactoring makes sense here.

bmermet replied:
It's not a matter of the change making sense on itself but to keep the size of the PR manageable and ease the review process. Separating functional changes from large refactors simplifies a lot the review (especially in a PR that has 3000 lines of changes). The [Google code review guide](https://github.com/google/eng-practices/blob/master/review/developer/small-cls.md#separate-out-refactorings-refactoring) makes a case for this for instance.

gbbr committed:
Move normalizer to pkg/trace/traceutil

gbbr committed:
Introduce new stats endpoint

gbbr committed:
Do not use ioutil.ReadAll and add tags to metrics

gbbr committed:
s/NewAggregationFromSpan/NewAggregation in doc

gbbr committed:
Add client_stats feature flag

gbbr committed:
Remove TopLevel field

bmermet commented on pkg/trace/api/api.go:
```diff
@@ -390,6 +402,38 @@ func (r *HTTPReceiver) rateLimited(n int64) bool {
+
+	ts := r.tagStats(v05, req)
+	rd := NewLimitedReader(req.Body, r.conf.MaxRequestBytes)
+	metrics.Count("datadog.trace_agent.receiver.stats_payload", 1, ts.AsTags(), 1)
```
Is there a reason why stats for handleStats are reported differently than in handleTraces?

gbbr replied:
Your questions is a bit ambiguous but I'll assume you refer to the atomic values. We use atomic values in hot paths which I think is not the case for either endpoint, especially not for this one.

bmermet replied:
Yes that's what I was referring to, thank you for the explanation. That's maybe work for another PR but I think we should make it consistent if there is no reason to have two different ways of reporting similar metrics.

bmermet commented on pkg/trace/api/api.go:
```diff
@@ -390,6 +402,38 @@ func (r *HTTPReceiver) rateLimited(n int64) bool {
+	ts := r.tagStats(v05, req)
+	rd := NewLimitedReader(req.Body, r.conf.MaxRequestBytes)
+	metrics.Count("datadog.trace_agent.receiver.stats_payload", 1, ts.AsTags(), 1)
+	if v := req.Header.Get("Content-Length"); v != "" {
```
Same thing, handleTraces uses `req.Body.(*LimitedReader).Count` to get the length of the payload, I think it would make sense to use the same approach in both functions.

gbbr replied:
Good point. Done.

bmermet commented on pkg/trace/api/api.go:
```diff
@@ -390,6 +402,38 @@ func (r *HTTPReceiver) rateLimited(n int64) bool {
+		httpDecodingError(err, []string{"handler:stats", "codec:msgpack", "v:v0.5"}, w)
+		return
+	}
+	metrics.Count("datadog.trace_agent.receiver.stats_buckets", int64(len(in.Stats)), ts.AsTags(), 1)
```
I can't figure out what writes the http response in this function, how does it work?

gbbr replied:
There is no need to, the server automatically responds 200 OK if the user doesn't ([example](https://play.golang.org/p/plxicRqw1-U))

gbbr committed:
Address comments

bmermet commented on pkg/trace/pb/stats.proto:
```diff
@@ -0,0 +1,37 @@
+syntax = "proto3";
```
Shouldn't the proto definition be removed to as per https://github.com/DataDog/datadog-agent/pull/6687#discussion_r537476515

gbbr committed:
Fix integration tests after removing TopLevel

bmermet commented on pkg/trace/obfuscate/obfuscate.go:
```diff
@@ -97,6 +98,22 @@ func (o *Obfuscator) Obfuscate(span *pb.Span) {
 
+// ObfuscateStatsGroup obfuscates the given stats bucket group.
+func (o *Obfuscator) ObfuscateStatsGroup(b *pb.ClientGroupedStats) {
+	switch b.Type {
```
Why is obfuscation only implemented for sql and redis?

gbbr replied:
Only those 2 obfuscators address the Resource, the others address specific tags on the Meta map.https://github.com/DataDog/datadog-agent/blob/0a4a94618d169aa09376261ecdba5677f65295a2/pkg/trace/obfuscate/obfuscate.go#L80-L98

bmermet replied:
Got it. Although "cassandra" is treated like "sql" in `Obfuscate` is there a reason to omit it here?

bmermet replied:
I added "cassandra", we can remove it later if it was a mistake.

gbbr replied:
I think you're right. Thanks for adding.

raphaelgavache committed:
Update field

raphaelgavache committed:
Fix tests

raphaelgavache committed:
Fix header

bmermet committed:
Add Cassandra to obfuscation

gbbr reviewed (commented):
🎉 Thanks for the help guys!

bmermet committed:
Reorder imports

gbbr commented on pkg/trace/api/api.go:
```diff
@@ -28,6 +28,8 @@ import (
 	"sync/atomic"
 	"time"
 
+	"github.com/tinylib/msgp/msgp"
```
I think you got it backwards 😅 I need to write a wiki for the coding style, otherwise I'm just annoying. I promise I will do it soon.

gbbr commented:
Looks like my test is flakey. The env. var. setting/unsetting in tests is a bad practice and I should find a better way to set feature flags in tests.

bmermet committed:
Use httptest instead of opening a local server for tests
